src/java.base/share/classes/jdk/internal/util/DecimalDigits.java

```diff
@@ -73,90 +74,179 @@ public final class DecimalDigits implements Digits {
     }
 
     /**
-     * Singleton instance of DecimalDigits.
+     * Constructor.
      */
-    public static final Digits INSTANCE = new DecimalDigits();
+    private DecimalDigits() {
+    }
 
     /**
-     * Constructor.
+     * For values from 0 to 99 return a short encoding a pair of ASCII-encoded digit characters in little-endian
+     * @param i value to convert
+     * @return a short encoding a pair of ASCII-encoded digit characters
      */
-    private DecimalDigits() {
+    public static short digitPair(int i) {
+        return DIGITS[i];
     }
 
-    @Override
-    public int digits(long value, byte[] buffer, int index,
-                      MethodHandle putCharMH) throws Throwable {
-        boolean negative = value < 0;
-        if (!negative) {
-            value = -value;
+    /**
+     * Returns the string representation size for a given int value.
+     *
+     * @param x int value
+     * @return string size
+     *
+     * @implNote There are other ways to compute this: e.g. binary search,
+     * but values are biased heavily towards zero, and therefore linear search
+     * wins. The iteration results are also routinely inlined in the generated
+     * code after loop unrolling.
+     */
+    public static int stringSize(int x) {
+        int d = 1;
+        if (x >= 0) {
+            d = 0;
+            x = -x;
         }
-
-        long q;
-        int r;
-        while (value <= Integer.MIN_VALUE) {
-            q = value / 100;
-            r = (int)((q * 100) - value);
-            value = q;
-            int digits = DIGITS[r];
-
-            putCharMH.invokeExact(buffer, --index, digits >> 8);
-            putCharMH.invokeExact(buffer, --index, digits & 0xFF);
+        int p = -10;
+        for (int i = 1; i < 10; i++) {
+            if (x > p)
+                return i + d;
+            p = 10 * p;
         }
+        return 10 + d;
+    }
```

Comment:
@wenshao How about of using [this](https://github.com/plokhotnyuk/jsoniter-scala/blob/6b72cf75ad7f53e8a285d512009d164c3eabbb3a/jsoniter-scala-core/jvm/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonWriter.scala#L2367-L2369) trick to avoid multiplications in a loop?

Reply:
This implementation is copied from Integer.stringSize. In 2015, @Shipilev modified the table lookup to the current implementation. In actual testing, this algorithm is faster.

Reply:
No doubt, but that algorithm could have been easily integrated into the Digits implementation without disruption.